core/src/main/scala/cats/Foldable.scala

```diff
@@ -215,6 +215,33 @@ import simulacrum.typeclass
       }
 
   /**
+   * Like `collectFirst` from `scala.collection.Traversable` but takes `A => Option[B]`
+   * instead of `PartialFunction`s.
```

Comment:
I think partial functions are very nice for this case. why not use them?

Reply:
For a `PartialFunction`, when `isDefinedAt` and `apply` relies on a shared computation, that computation will have to be repeated for the hit case? For example, giving a list of ids, you want to find the first id with which you can find a record in a DB. A `PartialFunction` would course one extra DB query. Also `Id => Option[Record]` is just very nature, making it a `PartialFunction` would be awkward right?

Reply:
:+1: Reviewing some old code, I've implemented it this way on a couple pre-cats projects.  I'd rather this suggestion than dropping it altogether.

Reply:
`TraversableOnce` does [an unspeakable thing](https://github.com/scala/scala/blob/v2.12.4/src/library/scala/collection/TraversableOnce.scala#L141-L148) to avoid the double evaluation.

Reply:
@rossabaker wow, that is unspeakable.  That certainly removed one of the incentives of adding this method. Just need a opposite version of `PartialFunction#lift`, maybe in mouse.

Reply:
[Function.unlift](https://www.scala-lang.org/api/2.12.4/scala/Function$.html#unlift[T,R](f:T=%3EOption[R]):PartialFunction[T,R])?

Reply:
done! closing this one.

Reply:
Oh, I thought you were talking about removing the incentive for `collectFst`.  What about `Foldable` without `collectFirst`?  I've needed it on `NonEmptyList` before.

Reply:
My mistake, my only incentive is for a A=>Option[B] collectFirst. I am reopening it. I wonder if we should add `collectFirst: PartialFunction[A, B]` and `collectFirstO: A => Option[B]`
For the default `collectFirst` implementation (which is going to be overriden for actual `TraversableOnce` data types)  shall we do the same trick as `TraversableOnce`?